test/focus-trap-react.test.js

```diff
@@ -317,6 +323,156 @@ describe('FocusTrap', () => {
         expect(document.body).toHaveFocus();
       });
     });
+
+    it('Does not call onPostActivate() until checkCanFocusTrap() has completed', async () => {
+      let hasActivated = false;
+      let hasRunPostActivate = false;
+      render(
+        <FocusTrapExample
+          focusTrapOptions={{
+            checkCanFocusTrap: () => pause(500),
+            onActivate: () => {
+              hasActivated = true;
+            },
+            onPostActivate: () => {
+              hasRunPostActivate = true;
+            },
+          }}
+        />
+      );
+
+      expect(hasActivated).toBe(false);
+      expect(hasRunPostActivate).toBe(false);
+
+      // Activate the focus trap
+      const activateTrapButton = screen.getByText('activate trap');
+      activateTrapButton.focus();
+      fireEvent.click(activateTrapButton);
+
+      expect(hasActivated).toBe(true);
+      expect(hasRunPostActivate).toBe(false);
+
+      await pause(5);
+
+      expect(hasActivated).toBe(true);
```

Comment:
Can this be combined into a `waitFor` statement? [waitFor](https://testing-library.com/docs/dom-testing-library/api-async#waitfor)'s default timeout is 1000ms so it should be more than enough.

```
waitFor(() => expect(hasActivated).toBe(true))
```

That would be more "in line" with testing with RTL.

Reply:
`waitFor` was the first thing I tried. `waitFor` has very different functionality though.

`waitFor` actually works like this:
1. Try calling the callback function immediately.
2. If it errors try again _immediately_
3. Keep trying over and over until it either stops throwing errors or the timeout time has been reached.

That is very different functionality to "wait x milliseconds" which is what the `pause()` function does.

Reply:
You're right, they are different, but why wouldn't `waitFor()` work? The assertion will fail when `waitFor()` does the immediate call, which is fine. Then 5ms will elapse and onActivate will get called, and then 45ms later, `waitFor()` will again try the assertion, and it should pass at that point. No?

Reply:
I need to be able to easily check that the function has been called when it is supposed to have be called. 

Importantly I also need to check that the function has _not_ been called when it should not have been called yet.

I need to:
check `false` > **wait** > check `true`.

`waitFor` is a loop that just does this:
check `true` > check `true` > check `true`

Reply:
Isn't that expected, though? Because focus-trap calls onActivate immediately on activation, then looks at checkCanFocusTrap, and then calls onPostActivate. 🤔

Reply:
This is a test. We want to test that the expected behaviour actually does what we expect it to do.

We don't want to make a change that causes `onPostActivate` to get called before `checkCanActivate` is resolved. We want to ensure that if that situation were to happen, our tests will fail so we can fix the bug before it gets released to the public.

Reply:
OK, there must be something I'm not understanding. You're closer to the test than I am, so I will trust your judgment. Thanks for the discussion; let's go with using pause() as you're doing it now. :)